content/cpi.md

```diff
@@ -0,0 +1,339 @@
+
+The process for constructing the instruction manually is similar to how we've done it client-side, except we'll be implementing it in Rust instead of Typescript now! As you can see from the code snippet above, the `Instruction` object contains the same information that we're used to - it still requires a `program_id`, vector of `AccountMeta` objects, and a byte buffer that represents the `instruction_data`.
+
+The `program_id` is just a Pubkey, so that should be pretty straightforward. The accounts vector on the other hand, will require us to make use of the [`vec`](https://doc.rust-lang.org/std/macro.vec.html) macro. The vec macro allows us to create a vector using array notation, like so:
+```rust
+let v = vec![1, 2, 3];
+assert_eq!(v[0], 1);
+assert_eq!(v[1], 2);
+assert_eq!(v[2], 3);
+```
+
+The `accounts` field of the `Instruction` struct is expecting a vector of [`AccountMeta`](https://docs.rs/solana-program/latest/solana_program/instruction/struct.AccountMeta.html) objects, which can be created one of two ways - either with `AccountMeta::new` or `AccountMeta::read_only`. Using the `new` constructor creates a metadata object for writable accounts, while the `read_only` constructor specifies that the account is not writable. Both constructors expect two parameters, `pubkey: Pubkey` and `is_signer: bool`. The account metadata struct returned from both looks like this, look familiar?
+```rust
+pub struct AccountMeta {
+    pub pubkey: Pubkey,
+    pub is_signer: bool,
+    pub is_writable: bool,
+}
+```
+An example of putting this in action may look something like
+```rust
+use solana_program::instruction::AccountMeta;
+
+vec![
+   AccountMeta::new(account1_pubkey, true),
+   AccountMeta::read_only(account2_pubkey, false),
+   AccountMeta::read_only(account3_pubkey, true),
+   AccountMeta::new(account4_pubkey, false),
+]
+```
+The final field of the instruction object is the data, as a byte buffer of course. You can create a byte buffer in Rust using the `vec` macro again, which has an implemented function allowing you to create a vector of certain length. Once you have initialized an empty vector, you would construct the byte buffer similar to before. The first element of the buffer should be an integer indicating which instruction you're targeting, the rest of the buffer will be for the serialized instruction data. Feel free to read up on some of the [features of the vec macro available to you here](https://doc.rust-lang.org/alloc/vec/struct.Vec.html#).
+```rust
+let mut vec = Vec::with_capacity(3);
+vec.push(1);
+vec.push(2);
+vec.extend_from_slice(&number_variable.to_le_bytes());
+```
+
```

Comment:
I refactored it to be two sections, "CPI with `invoke`" and "CPI with `invoke_signed`", but how do you feel about a third one just on manually building the `instruction` object? The instruction info applies to both `invoke` and `invoke_signed`, Thinking it would read like:

"CPI with `invoke`" 

"CPI with `invoke_signed`"

"Building the `Instruction`"